### PR TITLE
Ruleset propagation and matching logic

### PR DESCRIPTION
**Pull request by xichen2020**
cc @robskillington @cw9 @prateek @jeromefroe 

This PR adds logic to handle ruleset propagation, including watching and processing ruleset updates, as well as rule key changes. It then uses the rules to matching against metric IDs to determine the applicable policies.

There are a few interfaces (namely `BackendServer`, `Reporter`, and `Matcher`) that don't have concrete implementations yet. Just want to get them out there to get some early feedback. 

I'm going to work on adding concrete implementations for the matcher interface (e.g., caching), and hooking up the matcher to the reporter, which talks to the backend server that forwards metrics to the aggregation tier.

### CODE REVIEW

File: rules/rulesets.go [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+	defer r.Unlock()
+
+	if r.expiring {
+		return
+	}
+	now := r.nowFn()
+	if idx := r.findWithLock(now); idx > 0 {
+		n := copy(r.byCutoverAsc[0:], r.byCutoverAsc[idx:])
+		for i := n; i < len(r.byCutoverAsc); i++ {
+			r.byCutoverAsc[i] = nil
```
Comment (cw9) [inline rules/rulesets.go]: curious, why is this needed?

File: rules/namespaces.go [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+			if err := existing.MarkAlive(); err != nil {
+				n.log.WithFields(
+					xlog.NewLogField("ruleSetKey", existing.Key()),
+					xlog.NewLogErrField(err),
+				).Error("failed to mark ruleset as alive")
+			}
+		}
+	}
+
+	for new := range incoming {
```
Comment (jeromefroe) [inline rules/namespaces.go]: +1

File: rules/namespaces.go [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+	n := &namespaces{
+		key:          key,
+		store:        store,
+		opts:         opts,
+		ruleSetKeyFn: opts.RuleSetKeyFn(),
+		log:          opts.InstrumentOptions().Logger(),
+		proto:        &schema.Namespaces{},
+		rules:        make(map[string]*ruleSets),
+	}
+	n.runtimeValue = newRuntimeValue(key, store, n.toNamespaces, namespacesVersion, namespacesCutover, n.process, opts)
```
Comment (jeromefroe) [inline rules/namespaces.go]: I kind of like the user having to explicitly call `Start` themselves, but I don't feel strongly either way.

File: rules/rulesets.go [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+}
+
+// CanDelete returns true if the ruleset is tombstoned
+// and all the existing rules have expired
+func (r *ruleSets) CanDelete() bool {
+	r.RLock()
+	defer r.RUnlock()
+
+	tombstoned := r.liveStatus == ruleSetsTombstoned
+	expired := len(r.byCutoverAsc) == 0 ||
```
Comment (jeromefroe) [inline rules/rulesets.go]: +1

File: reporter/backend.go [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package reporter
+
+import "github.com/m3db/m3metrics/policy"
+
+// BackendServer is a backend server
+type BackendServer interface {
```
Comment (jeromefroe) [inline reporter/backend.go]: It isn't entirely clear to me what `BackendServer` will be used for in comparison to `Reporter`. Perhaps an additional comment here might be helpful?

File: rules/match.go [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+
+package rules
+
+import "github.com/m3db/m3metrics/policy"
+
+// Matcher matches rules against metric IDs
+type Matcher interface {
+	// Match matches rules against metric ID and returns
+	// applicable versioned policies
+	Match(id string) policy.VersionedPolicies
```
Comment (jeromefroe) [inline rules/match.go]: Does it make sense to make `id` a byte slice here? Just thinking that since the metric objects store a byte slice for the metric's id instead of a string, we may have to do a conversion to a string here.

File: reporter/backend.go (hunk 5 above)
Comment (xichen2020) [inline reporter/backend.go]: Sure thing, the idea is that the reporter will receive metrics coming from a snapshot, matching each metrics and finding the applicable policies, and forward the metric along w/ the policies to the backend server. The backend server is then responsible for sending the metric and the policies downstream, taking care of the appropriate data encoding and routing decisions. Will add clarifying comments.

File: rules/match.go (hunk 6 above)
Comment (xichen2020) [inline rules/match.go]: My gut feeling is that a string id is probably slightly easier to process in terms of applying filtering rules. Additionally, the metrics coming out of a snapshot are already keyed by string IDs, so it would require a string to bytes conversion if the API takes in a byte slice

File: rules/namespaces.go (hunk 3 above)
Comment (xichen2020) [inline rules/namespaces.go]: I'd prefer the users to explicitly call `Start`, which makes it more clear when the values start updating, and facilitates testing as well.

File: rules/namespaces.go (hunk 2 above)
Comment (xichen2020) [inline rules/namespaces.go]: Sure thing, can change to `newNs` perhaps

File: rules/rulesets.go (hunk 4 above)
Comment (xichen2020) [inline rules/rulesets.go]: Sure thing

File: rules/rulesets.go (hunk 1 above)
Comment (xichen2020) [inline rules/rulesets.go]: This is because even if the slice is shrunk, we are still holding reference to the entire underlying memory chunk, which means if the elements between `n` and `len(r.byCutoverAsc)` hold reference to some expensive nested objects, the garbage collector would need to spend a large amount of time sweeping them, which is certainly undesirable.

File: rules/value.go [hunk 7]
```diff
@@ -0,0 +1,333 @@
...
+}
+
+func (rv *runtimeValue) watchUpdates(watch kv.ValueWatch) {
+	defer rv.updateWg.Done()
+
+	for {
+		select {
+		case _, ok := <-watch.C():
+			if !ok {
+				return
```
Comment (xichen2020) [inline rules/value.go]: Hm why is this unexpected? In fact when a namespace is deleted and the corresponding ruleset is expired, we stop watching updates of that ruleset by closing the watch, and may start watching again if that namespace is resurrected later.

File: rules/value.go [hunk 8]
```diff
@@ -0,0 +1,333 @@
...
+		multiErr    = xerrors.NewMultiError()
+		historical  []interface{}
+	)
+
+	for ; currVersion > defaultVersion && currCutover.After(now); currVersion-- {
+		values, err := rv.store.History(rv.key, currVersion, currVersion+1)
+		if err != nil {
+			err = fmt.Errorf("error getting history for version %d: %v", currVersion, err)
+			multiErr = multiErr.Add(err)
+			continue
```
Comment (xichen2020) [inline rules/value.go]: I think so, this is so that if one version is corrupted in kv, we can still load the previous versions without preventing the entire set of historical versions from being loaded.

File: rules/match.go (hunk 6 above)
Comment (jeromefroe) [inline rules/match.go]: Agreed, and since a lot of our components (e.g. filtering) already expect a string it would require further changes elsewhere. We can perhaps revisit using a byte slice from end to end in the future.

File: rules/value.go (hunk 7 above)
Comment (cw9) [inline rules/value.go]: I see, sounds good

File: rules/cache/cache.go [hunk 9]
```diff
@@ -0,0 +1,328 @@
...
+	c.Lock()
+	if c.closed {
+		c.Unlock()
+		return errCacheClosed
+	}
+	c.closed = true
+	c.Unlock()
+
+	close(c.closedCh)
+	c.wgWorker.Wait()
```
Comment (cw9) [inline rules/cache/cache.go]: why do we need to wait here?

File: rules/cache/cache.go [hunk 10]
```diff
@@ -0,0 +1,389 @@
...
+
+func (c *cache) Set(namespace string, source Source) {
+	c.Lock()
+	results, exist := c.namespaces[namespace]
+	if !exist {
+		results = newResults(source)
+	} else {
+		results.source = source
+	}
+	c.namespaces[namespace] = results
```
Comment (cw9) [inline rules/cache/cache.go]: can we still add to a closed cache?

File: id/types.go [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+// THE SOFTWARE.
+
+package id
+
+import "crypto/md5"
+
+// TODO(xichen): move the hash type and function to m3x.
+
+// Hash represents a form of ID suitable to be used as map keys.
+type Hash [md5.Size]byte
```
Comment (robskillington) [inline id/types.go]: Can we move this to `m3x` and make `m3db` reference this package too?  

I feel like we need to do this now before we start depending on it.

File: rules/cache/options.go [hunk 12]
```diff
@@ -0,0 +1,185 @@
...
+const (
+	// InvalidateOne only invalidates a single invalid entry as needed.
+	InvalidateOne InvalidationMode = iota
+
+	// InvalidateAll invalidates all entries as long as one entry is invalid.
+	InvalidateAll
+)
+
+const (
+	defaultCapcity           = 200000
```
Comment (robskillington) [inline rules/cache/options.go]: Typo here, this should be `defaultCapacity`

File: rules/cache.go [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+	Match(id []byte) rules.MatchResult
+}
+
+// Cache caches the rule matching result associated with metrics.
+type Cache interface {
+	// Get returns the rule matching result associated with a metric id.
+	Get(namespace string, id []byte) rules.MatchResult
+
+	// Set sets the result source for a given namespace.
+	Set(namespace string, source Source)
```
Comment (robskillington) [inline rules/cache.go]: Would it not be slightly more idiomatic and matching the APIs of source to have the cache interface be more specific that `Get(...)` actually calls `Match(...)` and separate set/delete naming with something more like:
```go
type Cache interface {
  Match(namespace string, id []byte) rules.MatchResult
  RegisterSource(namespace string, source Source)
  UnregisterSource(namespace string)
  // ...
}
```

File: rules/cache/cache.go [hunk 14]
```diff
@@ -0,0 +1,378 @@
...
+	var res mrules.MatchResult
+	results, exists := c.namespaces[namespace]
+	if !exists {
+		return res, true
+	}
+	source := results.source
+	idHash := mid.HashFn(id)
+	elem, exists := results.elems[idHash]
+	if !exists {
+		if source == nil {
```
Comment (robskillington) [inline rules/cache/cache.go]: Isn't this basically a `panic(...)` situation?  Why would there ever be something that didn't come from a source?

Comment (robskillington) [inline rules/cache/cache.go]: Can we just not put this in `c.namespaces` if `source == nil`?

File: rules/cache/cache.go [hunk 15]
```diff
@@ -0,0 +1,378 @@
...
+	// NB(xichen): add pooling if deletion happens frequent enough.
+	toDelete := c.toDelete
+	c.toDelete = nil
+	c.Unlock()
+
+	deleted := 0
+	c.list.Lock()
+	for _, elems := range toDelete {
+		for id, elem := range elems {
+			delete(elems, id)
```
Comment (robskillington) [inline rules/cache/cache.go]: Is it worth deleting `id` from `elems` considering its unreachable after `doDelete(...)`?

File: rules/cache/cache.go [hunk 16]
```diff
@@ -0,0 +1,378 @@
...
+	evictCh    chan struct{}
+	deleteCh   chan struct{}
+	toDelete   []elemMap
+	wgWorker   sync.WaitGroup
+	closed     bool
+	closedCh   chan struct{}
+}
+
+// NewCache creates a new cache.
+func NewCache(opts Options) rules.Cache {
```
Comment (robskillington) [inline rules/cache/cache.go]: One thing I have a question about the entire Cache: is there a reason we give entries a TTL and then refresh them instead of simply evicting rules as the rules actually change due to explicit notifications from our KV watches?

File: rules/cache/cache.go [hunk 17]
```diff
@@ -0,0 +1,378 @@
...
+		}
+		// fall through if the cached result is invalid and we have a valid source.
+	}
+	if setType == dontSetIfNotFound {
+		return res, false
+	}
+	// NB(xichen): we are guaranteed to have a valid source when we get here, and
+	// the result is either not cached, or cached but invalid, in both cases we
+	// should use the source to compute the result and set it the cache if needed.
+	return c.setWithLock(namespace, id, idHash, results, exists), true
```
Comment (robskillington) [inline rules/cache/cache.go]: Can we simplify these logic gates?  Seems like a reasonable amount of of falling through/hard to follow paths.  Potentially some early exits might hurt with less indentation?

File: rules/cache/list.go [hunk 18]
```diff
@@ -0,0 +1,152 @@
...
+	idHash    id.Hash
+	result    rules.MatchResult
+	deleted   bool
+	expiryNs  int64
+	prev      *element
+	next      *element
+}
+
+func (e *element) ShouldExpire(now time.Time) bool {
+	return time.Unix(0, atomic.LoadInt64(&e.expiryNs)).Before(now)
```
Comment (robskillington) [inline rules/cache/list.go]: Might it be cheaper to do `atomic.LoadInt64(&e.expiryNs) < now.UnixNano()`?

File: rules/cache/list.go [hunk 19]
```diff
@@ -0,0 +1,152 @@
...
+}
+
+func (e *element) SetExpiry(t time.Time) {
+	atomic.StoreInt64(&e.expiryNs, t.UnixNano())
+}
+
+// list is a type-specific implementation of doubly linked lists consisting
+// of elements so when we retrieve match results, there is no conversion
+// between interface{} and the concrete result type.
```
Comment (robskillington) [inline rules/cache/list.go]: Yay Go....

File: rules/cache/list.go [hunk 20]
```diff
@@ -0,0 +1,152 @@
...
+type list struct {
+	head *element
+	tail *element
+	len  int
```
Comment (robskillington) [inline rules/cache/list.go]: Just to avoid needing to use a keyword for a field, perhaps `length` is better for this field?

File: rules/cache/list.go (hunk 19 above)
Comment (robskillington) [inline rules/cache/list.go]: Maybe we should think about generating these if we need to create another in the future.

File: rules/match.go [hunk 21]
```diff
@@ -0,0 +1,75 @@
...
+	return &matcher{
+		opts:        opts,
+		namespaceFn: opts.NamespaceFn(),
+		namespaces:  namespaces,
+		cache:       cache,
+	}, nil
+}
+
+func (m *matcher) Match(id []byte) rules.MatchResult {
+	namespace := m.namespaceFn(id)
```
Comment (robskillington) [inline rules/match.go]: As discussed we will need to change the interfaces here so that something like `ReportableMetric` can be queried for a tag, e.g. `TagValue(name string)`.  This way statsd and any othe rmetric being represented in the future can be matched into a namespace.

File: rules/namespaces.go [hunk 22]
```diff
@@ -0,0 +1,177 @@
...
+	// simply mark the rulesets for that namespace deleted and only
+	// remove it from the rules map when the rules have expired.
+	for ns, existing := range n.rules {
+		if _, ok := incoming[ns]; !ok {
+			if err := existing.MarkTombstoned(); err != nil {
+				n.log.WithFields(
+					xlog.NewLogField("ruleSetKey", existing.Key()),
+					xlog.NewLogErrField(err),
+				).Error("failed to mark ruleset as tombstoned")
+			}
```
Comment (robskillington) [inline rules/namespaces.go]: Perhaps a continue at the end of the `if` clause, to avoid the `else` clause and the indentation below.

File: rules/namespaces.go [hunk 23]
```diff
@@ -0,0 +1,177 @@
...
+			if err := existing.MarkAlive(); err != nil {
+				n.log.WithFields(
+					xlog.NewLogField("ruleSetKey", existing.Key()),
+					xlog.NewLogErrField(err),
+				).Error("failed to mark ruleset as alive")
+			}
+		}
+	}
+
+	for newNs := range incoming {
```
Comment (robskillington) [inline rules/namespaces.go]: Perhaps just `ns` instead of `newNs`?

File: rules/namespaces.go [hunk 24]
```diff
@@ -0,0 +1,177 @@
...
+		if _, ok := n.rules[newNs]; !ok {
```
Comment (robskillington) [inline rules/namespaces.go]:
```go
if _, ok := n.rules[newNs]; ok {
  continue
}

// inner
```
Perhaps?

File: rules/rulesets.go [hunk 25]
```diff
@@ -0,0 +1,263 @@
...
+	if r.expiring {
+		r.RUnlock()
+		return
+	}
+	r.RUnlock()
+
+	r.Lock()
+	defer r.Unlock()
+
+	if r.expiring {
```
Comment (robskillington) [inline rules/rulesets.go]: Seems like `r.expiring` is never set to true?  Is it useful at all?

File: runtime/value.go [hunk 26]
```diff
@@ -0,0 +1,355 @@
...
+
+	// The value may be closed at this point, in which case
+	// we don't override its status.
+	if v.status == valueStopping {
+		v.status = valueStopped
+	}
+	return nil
+}
+
+func (v *value) Close() error {
```
Comment (robskillington) [inline runtime/value.go]: Is there any reason why Stop and Close are different methods?  Do you ever Stop and not Close after?  i.e. it seems as soon as you are disinterested in the value you'd want to close up eventually anyhow no?

File: runtime/value.go [hunk 27]
```diff
@@ -0,0 +1,355 @@
...
+		currCutover = v.cutoverFn(latest)
+		multiErr    = xerrors.NewMultiError()
+		historical  []interface{}
+	)
+
+	for ; currVersion > defaultVersion && currCutover.After(now); currVersion-- {
+		values, err := v.store.History(v.key, currVersion, currVersion+1)
+		if err != nil {
+			err = fmt.Errorf("error getting history for version %d: %v", currVersion, err)
+			multiErr = multiErr.Add(err)
```
Comment (robskillington) [inline runtime/value.go]: Do we need retry here?  I worry that if there is ever an error we'll just always have skipped it and never recover from it for the lifetime of the entire collector (until it's restarted).

File: runtime/value.go [hunk 28]
```diff
@@ -0,0 +1,355 @@
...
+
+	// If there are multiple versions in between the cached version
+	// and the new version, we need to retrieve every version between
+	// the latest cached version and the new version.
+	multiErr := xerrors.NewMultiError()
+	if newVersion > cachedVersion+1 {
+		values, err := v.store.History(v.key, cachedVersion+1, newVersion)
+		if err != nil {
+			err = fmt.Errorf("error getting history from version %d to version %d: %v", cachedVersion+1, newVersion, err)
+			multiErr = multiErr.Add(err)
```
Comment (robskillington) [inline runtime/value.go]: Again: do we need retry here?  I worry that if there is ever an error we'll just always have skipped it and never recover from it for the lifetime of the entire collector (until it's restarted).

**Comment by robskillington** [comment]
Also another bigger question: how is this going to work at query time?  For each ID we are going to end up querying (for every ID in every query) do we need to run `Match(...)` for every single rule in the ruleset for that namespace?  

Seems potentially very expensive.

File: rules/cache/cache.go (hunk 14 above)
Comment (xichen2020) [inline rules/cache/cache.go]: Yeah, was originally thinking for namespaces without rules this would be useful, but seems we always create a new ruleset per namespace so that should be a non-issue. Will change.

File: rules/cache/cache.go (hunk 17 above)
Comment (xichen2020) [inline rules/cache/cache.go]: Yeah I tried to make the logic clear without duplicating code, can take another look to see if that can be further simplified.

File: rules/cache/cache.go (hunk 15 above)
Comment (xichen2020) [inline rules/cache/cache.go]: Can remove sure thing

File: rules/cache/list.go (hunk 18 above)
Comment (xichen2020) [inline rules/cache/list.go]: Yeah sounds good

File: rules/cache/list.go (hunk 19 above)
Comment (xichen2020) [inline rules/cache/list.go]: Yeah possibly. I had the same thoughts about the pools in m3x and talked to @kobolog about it a while back, guess that was put on the back burner atm. The general thoughts was having a template/ package that's used to generate type-specific types and have the go generate commands integrated w/ Makefile. There might be some minor issues with import paths etc, but overall felt doable. Could revisit this if handwriting basic data structures becomes a common occurrence.

File: rules/cache/list.go (hunk 20 above)
Comment (xichen2020) [inline rules/cache/list.go]: Sure thing

File: rules/cache/options.go (hunk 12 above)
Comment (xichen2020) [inline rules/cache/options.go]: Ah yeah good catch

File: rules/match.go (hunk 21 above)
Comment (xichen2020) [inline rules/match.go]: Yeah sounds good. That interface would also need to have an ID() method that returns the raw bytes.

File: rules/namespaces.go (hunk 22 above)
Comment (xichen2020) [inline rules/namespaces.go]: Sure

File: rules/namespaces.go (hunk 23 above)
Comment (xichen2020) [inline rules/namespaces.go]: yeah

File: rules/namespaces.go (hunk 24 above)
Comment (xichen2020) [inline rules/namespaces.go]: Sounds good

File: rules/rulesets.go (hunk 25 above)
Comment (xichen2020) [inline rules/rulesets.go]: Yeah I should've set it here, good catch, will add a test

File: runtime/value.go (hunk 26 above)
Comment (xichen2020) [inline runtime/value.go]: Yeah in the case where the rulesets are marked as tombstoned (e.g., when the corresponding namespace is deleted) but we haven't yet reached the cutover time of that change, we call `ruleSets.Stop()` to stop receiving updates but don't close the rulesets so that if the namespace is resurrected before cutover, we'll call `ruleSets.Start()` to start receiving updates again.

File: runtime/value.go (hunk 27 above)
Comment (xichen2020) [inline runtime/value.go]: Sure can add retry semantics here.

File: runtime/value.go (hunk 28 above)
Comment (xichen2020) [inline runtime/value.go]: Sure can add retry semantics here.

File: rules/rulesets.go (hunk 25 above)
Comment (xichen2020) [inline rules/rulesets.go]: On a second thought, it's not all that useful since we need a write lock when expiring so other goroutines will be blocked anyway, I'll just remove the flag.

File: runtime/value.go (hunk 26 above)
Comment (robskillington) [inline runtime/value.go]: I see, might it be worth renaming `Start` and `Stop` to `Watch` and `Unwatch` perhaps to make it a little more clear on the semantics of start/stopping is?

File: runtime/value.go (hunk 27 above)
Comment (robskillington) [inline runtime/value.go]: Is there a way to do indefinite background retries?  I feel like we should be able to support network partitions etc, and the only way to do that is indefinitely in the background be retrying until success.

File: runtime/value.go (hunk 26 above)
Comment (xichen2020) [inline runtime/value.go]: Sure yeah

File: runtime/value.go (hunk 27 above)
Comment (xichen2020) [inline runtime/value.go]: Yeah possibly, can take a stab at that.

File: rules/match.go [hunk 29]
```diff
@@ -21,46 +21,52 @@
...
 
-func (m *matcher) Match(id []byte) rules.MatchResult {
-	namespace := m.namespaceFn(id)
-	return m.cache.Get(namespace, id)
+func (m *matcher) Match(id metric.ID) rules.MatchResult {
+	ns, found := id.TagValue(m.namespaceTag)
+	if !found {
+		return m.cache.Match(m.defaultNamespace, id.Bytes())
+	}
+	return m.cache.Match(string(ns), id.Bytes())
```
Comment (robskillington) [inline rules/match.go]: Haha, ah yeah. Hm, is it worth making namespace all `[]byte` everywhere too?  Or at least in a follow up PR with a comment above this?

File: runtime/value.go [hunk 30]
```diff
@@ -251,105 +260,225 @@ func (v *value) watchUpdates(watch kv.ValueWatch) {
...
+			historical = append(historical, values[0])
+			currCutover = v.cutoverFn(values[0].value)
 			continue
 		}
-
-		if len(values) != 1 {
-			err = fmt.Errorf("expected 1 value for version %d but got %d", currVersion, len(values))
-			multiErr = multiErr.Add(err)
-			continue
+		if !xerrors.IsNonRetryableError(err) {
```
Comment (robskillington) [inline runtime/value.go]: There is a `xerrors.IsRetryableError(...)` to make this more readable.

File: rules/match.go (hunk 29 above)
Comment (xichen2020) [inline rules/match.go]: Done. All the namespaces are now `[]byte`.

**Comment by xichen2020** [comment]
@robskillington @cw9 @jeromefroe this is ready for review again. For recent changes, take a look at https://github.com/m3db/m3collector/compare/fbc078c...fcb87d7. Thanks!

File: rules/cache/list.go [hunk 31]
```diff
@@ -0,0 +1,152 @@
...
+	idHash   xid.Hash
+	result   rules.MatchResult
+	deleted  bool
+	expiryNs int64
+	prev     *element
+	next     *element
+}
+
+func (e *element) ShouldExpire(now time.Time) bool {
+	return atomic.LoadInt64(&e.expiryNs) <= now.UnixNano()
```
Comment (cw9) [inline rules/cache/list.go]: nit: add a test to guard the "=" case?

File: rules/namespaces.go [hunk 32]
```diff
@@ -0,0 +1,176 @@
...
+			latestSnapshot := snapshots[len(snapshots)-1]
+			// If the latest update shows the namespace is tombstoned, and we
+			// have received the corresponding ruleset update, we can stop watching
+			// the ruleset updates.
+			if latestSnapshot.Tombstoned() && latestSnapshot.ForRuleSetVersion() == ruleSet.Version() {
+				shouldWatch = false
+			}
+		}
+		if !shouldWatch {
+			ruleSet.Unwatch()
```
Comment (cw9) [inline rules/namespaces.go]: add a log here?

File: runtime/value.go [hunk 33]
```diff
@@ -0,0 +1,210 @@
...
+)
+
+const (
+	// DefaultVersion is the default version of an uninitialized kv value.
+	DefaultVersion = 0
+)
+
+var (
+	// ErrCreateWatch indicates an error when creating a watch.
+	ErrCreateWatch      = errors.New("error creating watch")
```
Comment (cw9) [inline runtime/value.go]: is this used?

File: runtime/value.go [hunk 34]
```diff
@@ -0,0 +1,210 @@
...
+		err = v.updateWithLockFn(watch.Get())
+	}
+
+	// NB(xichen): we want to start watching updates even though
+	// we may fail to initialize the value temporarily (e.g., during
+	// a network partition) so the value will be updated when the
+	// error condition is resolved.
+	go v.watchUpdates(v.watch)
+	if err != nil {
+		return InitValueError{innerError: err}
```
Comment (cw9) [inline runtime/value.go]: InitValueError was not checked outside this package? maybe just return err?

File: rules/cache/list.go (hunk 31 above)
Comment (xichen2020) [inline rules/cache/list.go]: Sure thing

File: rules/namespaces.go (hunk 32 above)
Comment (xichen2020) [inline rules/namespaces.go]: Hm not sure if I want to clutter the logs too much

File: runtime/value.go (hunk 33 above)
Comment (xichen2020) [inline runtime/value.go]: Yeah good catch, will remove

File: runtime/value.go (hunk 34 above)
Comment (xichen2020) [inline runtime/value.go]: While it's not strictly necessary, I can see the users check if the type of the error is `InitValueError` and take actions accordingly, instead of checking if the type of the error is not `CreateWatchError`, which I feel is more explicit, so I'd prefer to keep it unless you feel strongly.

Comment (cw9) [inline runtime/value.go]: yeah sounds reasonable
